kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k

```diff
@@ -57,31 +50,91 @@ module EVM-INT-SIMPLIFICATION-COMMON
   // #asWord
   // ###########################################################################
 
-  rule [asWord-lb]: 0 <=Int #asWord(_WS)     => true [simplification, smt-lemma]
-  rule [asWord-ub]: #asWord(_WS) <Int pow256 => true [simplification, smt-lemma]
-
-  rule [asWord-lt]:
-    #asWord ( BA ) <Int X => true
-    requires 2 ^Int ( 8 *Int lengthBytes(BA) ) <=Int X
-    [concrete(X), simplification]
-
-  rule [asWord-lt-concat]:
-    #asWord ( BA1 +Bytes BA2 ) <Int X => #asWord ( BA1 ) <Int ( X -Int #asWord ( BA2 ) ) /Int ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
-    requires ( X -Int #asWord ( BA2 ) ) modInt ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) ) ==Int 0
-    [concrete(BA2, X), simplification, preserves-definedness]
-
-  rule [asWord-lt-range]:
-    #asWord ( #range ( _:Bytes, S, W ) ) <Int X => true
-    requires 0 <=Int S andBool 0 <=Int W
-     andBool 2 ^Int ( 8 *Int W ) <=Int X
-    [simplification, concrete(S, W, X), preserves-definedness]
-
-  rule [lt-asWord-range]:
-    X <Int #asWord ( #range ( B:Bytes, S, W ) ) => X <Int #asWord ( B )
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord]:
+    #asWord ( B1 ) ==Int #asWord ( B2 ) => #buf ( 32 -Int lengthBytes(B1), 0 ) +Bytes B1 ==K #buf ( 32 -Int lengthBytes(B2), 0 ) +Bytes B2
+    requires lengthBytes(B1) <=Int 32 andBool lengthBytes(B2) <=Int 32
+    [simplification, preserves-definedness]
+
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord-ml]:
+    { #asWord ( B1 ) #Equals #asWord ( B2 ) } => { true #Equals #asWord ( B1 ) ==Int #asWord ( B2 ) }
+    [simplification, preserves-definedness]
+
+  // Unification: `#asWord` and concrete number
+  rule [asWord-eq-num]:
+    #asWord ( B:Bytes ) ==Int X:Int =>
+      0 <=Int X andBool X <Int 2 ^Int (8 *Int minInt(32, lengthBytes(B)))
+                andBool (
+                  ( lengthBytes(B) <=Int 32 andBool #buf ( 32 -Int lengthBytes(B), 0 ) +Bytes B ==K #buf ( 32, X ) )
+                    orBool
+                  ( lengthBytes(B) >Int 32 andBool #range ( B, lengthBytes(B) -Int 32, 32 ) ==K #buf ( 32, X ) )
+                )
+    [simplification, concrete(X), comm, preserves-definedness]
```

Comment:
Can we write this as multiple rules instead? The first clause can be a `requires` clause of both rules, and then each rule can make a choice based on `lengthBytes(B) <=Int 32` basically.

Are there cases where we cannot tell which of the cases in the `orBool` we fall into, but we can tell that _one_ of the cases applies?

Reply:
I also wonder whether this lemma is needed, and if so do we have tests of it directly here? It's by far the most "expansive" in the sense that the RHS is much larger than the LHS (in term size) of the whole PR, so it doesn't necessarily feel like it's a simplificaiton.

Reply:
Ah, I see `concrete(X)`, which means the RHS will be simpler by a decent amount. I still think the first condition can definitely by a side-condition, with a branch of rules:

```
rule #asWord(B) ==Int X => false requires notBool #rangeUInt(256, X)
rule #asWord(B) ==Int X => #buf(32 -Int lengthBytes(B), 0) +Bytes B ==K #buf(32, X) requires #rangeUInt(256, X) andBool lengthBytes(B) <=Int 32
rule #asWord(B) ==Int X => #range(B, lengthBytes(B) -Int 32, 32) ==K #buf(32, X) requires #rangeUInt(256, X) andBool 32 <Int lengthBytes(B)
```

I guess in general we should have pretty good knowledge of `B`, so it should be able to pick which rule applies here. Are there cases where we don't have that knowledge? Maybe when processing dynamically sized calldata? Do we have a test here of that case?

Reply:
Yes, we should know the length of B, so we should get one of the branches only. Happy to change.

Reply:
I think that we should always be able to tell whether the length is smaller or greater than 32 in the context of asking for an `#asWord`, I don't see why the compiler why try to convert arbitrarily-sized dynamic data to a single word.

Reply:
Ah, there's a problem. If we really can't choose a branch, then we would have to change both of these to each have three cases as well:
```
  rule [asWord-eq-num-ml-l]: { #asWord ( B:Bytes ) #Equals X:Int } => { true #Equals #asWord ( B:Bytes ) ==K X:Int } [simplification, concrete(X), preserves-definedness]
  rule [asWord-eq-num-ml-r]: { X:Int #Equals #asWord ( B:Bytes ) } => { true #Equals #asWord ( B:Bytes ) ==K X:Int } [simplification, concrete(X), preserves-definedness]
```
because otherwise the simplification would just go back and forth between the two.

Reply:
Re tests, yes, there are four, exercising all branches. They are starting from
https://github.com/runtimeverification/evm-semantics/blob/e73f8d93607156da28d22ea63b6e0f42c890726a/tests/specs/functional/lemmas-spec.k#L670

Reply:
In summary for this one, I think that either it needs to stay as-is, knowing that the worst that can happen is that we get both branches on `lengthBytes(B)`, which, from my experience, is not likely, or we duplicate the reasoning for the ML-level lemmas.

Reply:
Does using rule priorities make the simplificaitons non-looping?

Reply:
Not in this case, no.